Add output in JSON for `solana-ledger-tool bounds` subcommand

gnapoli23 opened the pull request:
#### Problem
When using `solana-ledger-tool` the only possible output in plaintext.

#### Summary of Changes
- Add a struct to capture the relevant information (slots and roots)
- Use serde::json for json, and a custom `Display` trait implementation for the regular text output

Note: This PR changes the regular output format for ledgers that have either no slots at all, or some slots but no roots. See specifics [here](https://github.com/solana-labs/solana/pull/28410#pullrequestreview-1246780370) in here as well as further messages for why the breakage was justified.

Fixes #24756

steviez commented:
@gnapoli23 - Sorry about the delay on this PR. Would you mind rebasing on tip-of-master and resolving the conflict that the GUI says is present? Can try to take a look in the next couple days once you have done that

gnapoli23 commented:
@steviez Sure, i'll rebase asap

CriesofCarrots commented:
I don't really love a new `bounds.rs` module, since that subcommand is a rather simple one, and doesn't support any lower subcommands, the way `bigtable` does. What about instead adding an `output.rs` module to hold the new output structs, and otherwise handling the logic in situ until such time as we do a larger solana-ledger-tool refactor?

CriesofCarrots commented:
Try `git pull --rebase` from upstream master, and then force push to your branch.

gnapoli23 commented:
@CriesofCarrots Yeah i can surely roll the changes back in situ and add only an `output.rs` module. To be honest I followed the `bigtable.rs` structuring because i thought that was your idea for refactoring the crate. Thanks for the suggestion, i'll work on it 👍🏻

steviez commented on ledger-tool/src/output.rs:
```diff
@@ -0,0 +1,51 @@
+
+#[derive(Serialize, Debug, Default)]
+#[serde(rename_all = "camelCase")]
+pub struct SlotBounds {
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub total_slots: Option<SlotInfo>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub rooted_slots: Option<SlotInfo>,
+}
```
I'm honestly not sure of which of these is preferred for the empty ledger case:
```
{}
```
vs.
```
{
  "totalSlot": null,
  "rootedSlots": null
}
```
Will circle back on this unless Tyera chimes in with any opinions; I was probably going to look at some of our RPC methods to see what convention we use there

CriesofCarrots replied:
I actually think something like this would be more clear (using my proposed field-name changes):
```
{
    slots: {
        total: 0,
        first: null,
        last: null,
        num_after_last_root: 0
    },
    roots: {
        total: 0,
        first: null,
        last: null
    }
}
```
Will require making a couple more `SlotInfo` fields Options.

steviez replied:
+1 for above

steviez reviewed (commented):
Looking pretty good; took a quick pass and will take another one soon

gnapoli23 commented on ledger-tool/src/main.rs:
```diff
@@ -4185,55 +4190,104 @@ fn main() {
-                                    println!("Non-empty slots: {slots:?}");
+
+                            slot_bounds.total_slots = Some(SlotInfo {
+                                start: *first,
+                                end: *last,
```
Would it be ok for `end`?

`end: *slots.last().unwrap_or(slots.first().unwrap())`

This should cover the case in which `start` and `end` have the same value (e.g. initial case)

steviez replied:
The pre-existing use of `unwrap_or()` was unnecessary.
- We already know that `slots` has at least one element, so the regular `.unwrap()`'s are ok
- For the case where `slots` has exactly one element, `.first()` and `.last()` will return the same element.

If you'd like to prove it to yourself, add and run the following unit test with `cargo test test_single_element_vector_sanity`:
```rust
    #[test]
    fn test_single_element_vector_sanity() {
        // Create vector with single element (3)
        let v = vec![3; 1];
        assert_eq!(v.len(), 1);
        assert_eq!(v.first().unwrap(), v.last().unwrap());
        assert_eq!(*v.first().unwrap(), 3);
    }
```

gnapoli23 replied:
Perfectly clear, thanks. I'll update it as suggested :)

CriesofCarrots commented on ledger-tool/src/main.rs:
```diff
@@ -4185,55 +4190,104 @@ fn main() {
+                        let mut slot_bounds = SlotBounds::default();
+
                         if slots.is_empty() {
-                            println!("Ledger is empty");
+                            match output_format {
+                                OutputFormat::Json | OutputFormat::JsonCompact => {
+                                    println!("{}", output_format.formatted_string(&slot_bounds))
+                                }
+                                _ => println!("Ledger is empty"),
+                            }
```
I agree that we don't need to inspect root data if there are zero slots. But I think the right answer is to populate `SlotBounds` here, and only have one print statement. Something like:
```
let slot_bounds = if slots.is_empty() {
    SlotBounds::default()
} else {
    ... // existing stuff, plus rooted_slot_iterator handling
};
println!("{}", output_format.formatted_string(&slot_bounds));
```

steviez replied:
+1 for above suggestion

gnapoli23 replied:
This was my first approach to it, first collecting info and then printing by considering the output format but didn't want to change things too much. I'll follow this path, thanks! 🙏

CriesofCarrots commented on ledger-tool/src/output.rs:
```diff
@@ -0,0 +1,51 @@
+#[serde(rename_all = "camelCase")]
+pub struct SlotBounds {
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub total_slots: Option<SlotInfo>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub rooted_slots: Option<SlotInfo>,
```
```suggestion
    pub slots: Option<SlotInfo>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub roots: Option<SlotInfo>,
```

CriesofCarrots commented on ledger-tool/src/main.rs:
```diff
@@ -4185,55 +4190,104 @@ fn main() {
-                                    println!("Non-empty slots: {slots:?}");
+
+                            slot_bounds.total_slots = Some(SlotInfo {
+                                start: *first,
+                                end: *last,
```
"Start" and "end" are not intuitive for me. How about "first/last" or "lowest/highest"?

gnapoli23 replied:
First/last are great, updating 👍🏻

CriesofCarrots commented on ledger-tool/src/output.rs:
```diff
@@ -0,0 +1,51 @@
+    pub end: u64,
+    pub total: usize,
+    #[serde(rename = "fromLast", skip_serializing_if = "Option::is_none")]
+    pub from_last: Option<u64>,
```
```suggestion
    pub num_after_last_root: Option<u64>,
```

CriesofCarrots commented on ledger-tool/src/output.rs:
```diff
@@ -0,0 +1,51 @@
+    pub from_last: Option<u64>,
+}
+
+impl Display for SlotInfo {
```
I'm not very clear on what these Display implementations are for, since we're not using them in `main.rs`. Am I missing something?

gnapoli23 replied:
No it's related to previous implementation based on `bounds.rs` module, following `bigtable.rs` subcommand as reference; i tought it was the idea to refactor `ledger-tool` crate, so i followed it :)
Considering what you suggested after reviewing sources, i guess that implementing the `Display` trait is ok but its logic should be the one used for "simple text printing". Did i understand correctly? Thank you :)

steviez replied:
> i guess that implementing the Display trait is ok but its logic should be the one used for "simple text printing"

Yes, implementing the Display trait is fine; however, CriesofCarrots comment is pointing out that as you have it, you are not using utilizing the Display trait implementation. Because we are implementing Display trait, this
```rust
                        match output_format {
                            OutputFormat::Json | OutputFormat::JsonCompact => {
                                println!("{}", output_format.formatted_string(&slot_bounds))
                            }
                            _ => {
                                // Simple text print in other cases.
                                ...
```
should just become
```rust
                        println!("{}", output_format.formatted_string(&slot_bounds))
```

gnapoli23 replied:
Got it, i'll update as suggested so. Thanks!

CriesofCarrots reviewed (commented):
I took a pass. Will take another after current comments are addressed. Thanks for your work on this!

CriesofCarrots commented:
Also, a git request: could you please eliminate the merge commits by rebasing instead? 🙏

steviez commented on ledger-tool/src/main.rs:
```diff
@@ -4185,55 +4189,140 @@ fn main() {
+                            bounds
+                        };
+
+                        // if slots.is_empty() {
```
Just incase you didn't know, you can do block comments like this:
```rust
/*
all
of
these
lines
will
be
commented
*/
```
Less tedious than commenting out each individual line

gnapoli23 replied:
Oh sure, i was just using VS code shortcuts for commenting (ctrl+k+c) :)

steviez replied:
Ha nice, figured as much

steviez commented on ledger-tool/src/main.rs:
```diff
@@ -4185,55 +4189,140 @@ fn main() {
+
+                                let count_past_root = slots.iter()
+                                    .rev()
+                                    .take_while(|slot| *slot > &last_rooted)
```
Not sure that the `&` is necessary here

gnapoli23 replied:
Yes it is needed, 'cause the closure takes a `slot` that is a `&&u64`, so the compiler suggested

steviez replied:
Yeah, it wasn't immediately obvious to me that within the closure `slot` would be `&&u64`, but I should have checked [the docs](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.take_while) as they explicitly state it out.

> Because the closure passed to take_while() takes a reference, and many iterators iterate over references, this leads to a possibly confusing situation, where the type of the closure is a double reference:

steviez commented on ledger-tool/src/output.rs:
```diff
@@ -0,0 +1,56 @@
+#[derive(Serialize, Debug, Default)]
+#[serde(rename_all = "camelCase")]
+pub struct SlotBounds {
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub slots: Option<SlotInfo>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub roots: Option<SlotInfo>,
```
To match the suggestion made elsewhere to make the first/last fields options, these should no longer be options
```suggestion
    pub slots: SlotInfo,
    pub roots: SlotInfo,
```

gnapoli23 replied:
About this, is the logic to print info there's also an `all` parameter coming from `matches` took from command line options. This parameter has the following purpose:
```
if all {
    println!("Non-empty slots: {:?}", slots);
}
```
Is it ok to add another parameter (a `bool` for instance) inside `SlotBounds` in order to accomplish this logic?

CriesofCarrots replied:
Yes, you can add a parameter tagged with either `#[serde(skip_serializing)]` or `#[serde(skip_serializing_if = "Option::is_none")]`, depending on what kind of field you need, to pass data to use in the Display implementation, without returning that data in the JSON map.

gnapoli23 commented:
I've updated sources, but have few more questions:

- if `Display` implementation for `SlotBounds`, should `Display` trait impl for `SlotInfo` be removed/updated?
- i passed another parameter for `all` case mentioned before, which is a reference to `slots` present in `main.rs`. If this `Option` field is not `None` than it means that all slots should be print, otherwise not. Is it ok?
- in `SlotBounds` i set `roots` as `Option<SlotInfo>` while leaving `slots` as is, considering  that rooted slots may indeed be optional. If not ok, i'll remove the `Option`. Thanks :)

steviez commented:
> * if `Display` implementation for `SlotBounds` is ok, should `Display` trait impl for `SlotInfo` be removed/updated?

I think we can remove the `Display` impl for `SlotInfo`. Given that these structs are exclusively for `bounds` subcommand and that we'll be outputting `SlotBounds` (and not `SlotInfo`), we won't be using `SlotInfo`'s `Display` impl anywhere. `SlotInfo` just saves us some repeated logic/fields in impl of `SlotBounds`, but don't think it needs to be a fully "standalone" struct.

> * i passed another parameter for `all` case mentioned before, which is a reference to `slots` present in `main.rs`. If this `Option` field is not `None` than it means that all slots should be print, otherwise not. Is it ok?

I believe the question you're asking is whether we should print `all_slots: null` or not print `all_slots` in the `all_slots == None` case, right?

Assuming so, we decided to print the other fields, even if they were `null`. However, that creates an interesting scenario for the empty ledger case:
- With empty ledger, we make `SlotBounds::default()`, which puts `all_slots` at `None`
- If the user has an empty ledger and passes `--all`, it could be argued that they'd expect to see an empty array as the value for `all_slots` all opposed to the `null` they'd get as we have it coded now.

Lemme think on this one for a little and/or let's see if CriesofCarrots has any opinion

> * in `SlotBounds` i set `roots` as `Option<SlotInfo>` while leaving `slots` as is, considering  that rooted slots may indeed be optional. If not ok, i'll remove the `Option`.

Per https://github.com/solana-labs/solana/pull/28410#discussion_r1055156135, lets make `roots` a regular `SlotInfo`, not an `Option<>`.

> Thanks :)

No, thank you! We appreciate you working on this and working through all the requests we've made, and I think we're pretty close to having this good-to-go.

gnapoli23 commented:
> I think we can remove the `Display` impl for `SlotInfo`. Given that these structs are exclusively for `bounds` subcommand and that we'll be outputting `SlotBounds` (and not `SlotInfo`), we won't be using `SlotInfo`'s `Display` impl anywhere. `SlotInfo` just saves us some repeated logic/fields in impl of `SlotBounds`, but don't think it needs to be a fully "standalone" struct.

Got it, so removing that `impl Display for SlotInfo` is ok, btw if you think that some logic could be "delegated" to that implementation, i'll update it.

> I believe the question you're asking is whether we should print `all_slots: null` or not print `all_slots` in the `all_slots == None` case, right?
> 
> Assuming so, we decided to print the other fields, even if they were `null`. However, that creates an interesting scenario for the empty ledger case:
> 
> * With empty ledger, we make `SlotBounds::default()`, which puts `all_slots` at `None`
> * If the user has an empty ledger and passes `--all`, it could be argued that they'd expect to see an empty array as the value for `all_slots` all opposed to the `null` they'd get as we have it coded now.
>
> Lemme think on this one for a little and/or let's see if CriesofCarrots has any opinion

Yes that's right, i was thinking at this specific case. Moreover, my idea was to use `all_slots` as a parameter to discriminate somehow in the printing logic if we should print also `all` the slots, as requested through the cli. `all_slots` right now holds the reference to the entire `Vec<u64>` of slots; it is also an `Option`, and:

- if `all_slots == None` it means that the user didn't pass `--all` as parameter through cli, so only "default" printing about slots and roots should be done;
- if `all_slots != None`, it means that we have `--all` as parameter, and for this reason we have to display also this info.

> Per [#28410 (comment)](https://github.com/solana-labs/solana/pull/28410#discussion_r1055156135), lets make `roots` a regular `SlotInfo`, not an `Option<>`.

Ok, i'll put it back as non-`Option` so 👍🏻 

> No, thank you! We appreciate you working on this and working through all the requests we've made, and I think we're pretty close to having this good-to-go.

It's a pleasure to contribute! 😄 
I would also take this opportunity to wish you guys a Merry Christmas and a Happy New Year!

CriesofCarrots commented on ledger-tool/src/main.rs:
```diff
@@ -4185,59 +4189,66 @@ fn main() {
-                        let all = arg_matches.is_present("all");
+                        let output_format =
+                            OutputFormat::from_matches(arg_matches, "output_format", false);
+                        let all = matches.is_present("all");
```
```suggestion
                        let all = arg_matches.is_present("all");
```

gnapoli23 replied:
Updated

gnapoli23 commented on ledger-tool/src/main.rs:
```diff
@@ -4185,59 +4189,66 @@ fn main() {
+                                    last: Some(*slots.last().unwrap()),
+                                    ..Default::default()
+                                },
+                                ..Default::default()
```
Fixed

gnapoli23 commented on ledger-tool/src/main.rs:
```diff
@@ -4185,59 +4189,66 @@ fn main() {
                     }
                     Err(err) => {
-                        eprintln!("Unable to read the Ledger: {err:?}");
+                        eprintln!("Unable to read the Ledger: {:?}", err);
```
Yes it came from the rebase, fixed by `git pull --rebase` on last commits

steviez replied:
```suggestion
                        eprintln!("Unable to read the Ledger: {err:?}");
```

gnapoli23 commented on ledger-tool/src/output.rs:
```diff
@@ -0,0 +1,81 @@
+    pub num_after_last_root: Option<usize>,
+}
+
+impl Display for SlotInfo {
```
Removed

gnapoli23 commented on ledger-tool/src/output.rs:
```diff
@@ -0,0 +1,81 @@
+impl VerboseDisplay for SlotBounds<'_> {}
+impl QuietDisplay for SlotBounds<'_> {}
+
+impl Display for SlotBounds<'_> {
```
Added logic as suggested. For the root Option fields, i had to leave the `if let Some(...)` logic, 'cause there was a `move` issue for `slots`.
By using `as_ref` on the option field
`let roots = self.roots.as_ref().unwrap_or_default()` i had 
`the trait bound &output::SlotInfo: std::default::Default is not satisfied
the trait std::default::Default is implemented for output::SlotInfo`

I could use a workaround such as
`let roots = self.roots.as_ref().unwrap_or(&SlotInfo::default())`
but didn't seem a clean way to do, so i preferred to leave it as was.

Let me know if you have any other suggestion, thanks.

steviez replied:
From another comment (linked below), we want `roots` to be a `SlotInfo`, not an `Option<SlotInfo>`
https://github.com/solana-labs/solana/pull/28410#discussion_r1055156135

Doing that should should remove some of the issues you talked about such as using `as_ref()`, so let's see how things look once `roots` is no longer an `Option`

steviez reviewed (commented):
Just posting these for reference so you can check; here is what the current output looks like for:
```
// Empty ledger
Ledger is empty
  with 0 rooted slots from 0 to 0
  and 0 slots past the last root
  
// No roots
Ledger has data for 11 slots 171368538 to 171368548
  with 0 rooted slots from 0 to 0
  and 11 slots past the last root

// Some roots
Ledger has data for 188 slots 171368538 to 171368725
  with 2 rooted slots from 171368539 to 171368540
  and 185 slots past the last root
```
As you have it implemented now, we'll just output "Ledger is empty" for the empty ledger case; I'm good with that as the lines that say no slots / roots are redundant at that point.

gnapoli23 commented:
@steviez Ok i think i fixed the rebase issue, i've also `reword`ed git messages by adding the issue number (#24756) to each one. Than i removed the Option for `SlotBounds` roots field and updated logic 👍🏻

steviez commented on ledger-tool/src/main.rs:
```diff
@@ -4167,59 +4171,66 @@ fn main() {
+
+                            // Consider also rooted slots, if present
+                            if let Ok(rooted) = blockstore.rooted_slot_iterator(0) {
+                                let mut first_rooted = 0;
+                                let mut last_rooted = 0;
+                                let mut total_rooted = 0;
+                                for (i, slot) in rooted.into_iter().enumerate() {
+                                    if i == 0 {
+                                        first_rooted = slot;
+                                    }
+                                    last_rooted = slot;
+                                    total_rooted += 1;
                                 }
+
+                                let count_past_root = slots
+                                    .iter()
+                                    .rev()
+                                    .take_while(|slot| *slot > &last_rooted)
+                                    .count();
```
```suggestion
                                let mut first_rooted = None;
                                let mut last_rooted = None;
                                let mut total_rooted = 0;
                                for (i, slot) in rooted.into_iter().enumerate() {
                                    if i == 0 {
                                        first_rooted = Some(slot);
                                    }
                                    last_rooted = Some(slot);
                                    total_rooted += 1;
                                }

                                let last_root_for_comparison = last_rooted.unwrap_or_default();
                                let count_past_root = slots
                                    .iter()
                                    .rev()
                                    .take_while(|slot| *slot > &last_root_for_comparison)
                                    .count();
```
Prior to this change, we might get output like:
```
{
  "slots": {
    "total": 101,
    "first": 171368538,
    "last": 171368638
  },
  "roots": {
    "total": 0,
    "first": 0,
    "last": 0,
    "numAfterLastRoot": 101
  }
}
```
Notice that the first/last root are both 0; this is because we were setting them to Some(0) even if there weren't any roots. So, this change will make the json output display `null` if there are no roots:
```
{
  "slots": {
    "total": 101,
    "first": 171368538,
    "last": 171368638
  },
  "roots": {
    "total": 0,
    "first": null,
    "last": null,
    "numAfterLastRoot": 101
  }
}
```

gnapoli23 replied:
That's perfectly clear :) I'll update right now

steviez reviewed (commented):
I think just my last two comments, and then we should be good to run through CI again

gnapoli23 commented:
@steviez Just uypdated 👍🏻

steviez commented:
> @steviez Just updated 👍🏻

Cool, looks like you got everything; started CI

gnapoli23 commented:
@steviez I rebased to pull latest master branch updates

CriesofCarrots reviewed (commented):
Nothing else from me, but I want to point out that we are making one technically breaking behavior change here, to confirm you are good with that @steviez ?

The different behavior occurs for a ledger with no roots (which wouldn't really occur in the wild, only if you purge all the roots from an existing ledger)...

Before:
```
$ solana-ledger-tool bounds
...
Ledger has data for 9 slots 1 to 9
  with 0 rooted slots from 0 to 0
  and 9 slots past the last root
```

With this change:
```
$ cargo run -- bounds
...
Ledger has data for 9 slots 1 to 9
  with no rooted slots

```

...or when the ledger is completely empty

Before:
```
$ solana-ledger-tool bounds
...
Ledger is empty
  with 0 rooted slots from 0 to 0
  and 0 slots past the last root
```

With this change:
```
$ cargo run -- bounds
...
Ledger is empty

```

steviez commented:
> Nothing else from me, but I want to point out that we are making one technically breaking behavior change here, to confirm you are good with that @steviez ?

Good callout and yeah, I'm good with it. Given that the command previously printed human-friendly output but not automation friendly output, I'm inclined to think not too many people would have been depending on this. Although in fairness, it wouldn't have been a terrible regex to parse.

> The different behavior occurs for a ledger with no roots (which wouldn't really occur in the wild, only if you purge all the roots from an existing ledger)...
> 
> Before:
> 
> ```
> $ solana-ledger-tool bounds
> ...
> Ledger has data for 9 slots 1 to 9
>   with 0 rooted slots from 0 to 0
>   and 9 slots past the last root
> ```
> 
> With this change:
> 
> ```
> $ cargo run -- bounds
> ...
> Ledger has data for 9 slots 1 to 9
>   with no rooted slots
> ```

Arguably, the old behavior for no roots is flawed / misleading:
- `with 0 rooted slots` would indicate no roots, while `from 0 to 0` could be interpreted as an inclusive range such as `[0, 0]` which has exactly one element, `0`
- `9 slots past the last root` is misleading for the same reason as it also implies that we have a root that is smaller than the lowest slot

> ...or when the ledger is completely empty
> 
> Before:
> 
> ```
> $ solana-ledger-tool bounds
> ...
> Ledger is empty
>   with 0 rooted slots from 0 to 0
>   and 0 slots past the last root
> ```
> 
> With this change:
> 
> ```
> $ cargo run -- bounds
> ...
> Ledger is empty
> ```

Looking at the two lines that we remove in this scenario,
- As mentioned above, I'm not fond of `with 0 rooted slot from 0 to 0`
- `with 0 slots past the last root` is true in this case, but also redundant knowing that we don't have any slots at all

I guess there are two questions:
1) Do we think the new output is better / more correct than the old format?
2) Are we worried that changing the output will break someone's automation?

I think you can tell where I stand on 1. , but open to hearing any counterpoints. As for 2., the GH issue that this PR would close didn't have much traffic / people asking for it. 

I think I've made my stance on 1 clear. As for 2., I'm glad we did this change and think it can enable some automation on our side, but we also didn't have a ton of people kicking our doors down for this either. We're changing the output, but we're also providing a better alternative for someone who want to parse in automated fashion.

That being said, you've been around longer and have probably seen a lot more sh*t than me, so maybe I'm being overzealous. Was your prompt to ensure that we're on the same page about altering output, or do you have a major concern about doing so such that you'd like to restore equivalent behavior for the non-json case?

CriesofCarrots reviewed (approved):
> Was your prompt to ensure that we're on the same page about altering output, or do you have a major concern about doing so such that you'd like to restore equivalent behavior for the non-json case?

The former. We need to be very intentional anytime we break backward compatibility; wanted to make sure that wasn't an oversight. But I am ultimately okay with this breakage, since it only affects severely pruned or empty ledgers.

steviez commented:
>  We need to be very intentional anytime we break backward compatibility; wanted to make sure that wasn't an oversight.

Good deal and yup, called it out in an [earlier comment](https://github.com/solana-labs/solana/pull/28410#pullrequestreview-1241644971), although re-reading my comment, your comment that clearly shows before & after is much more obvious.

steviez commented:
@gnapoli23 - CriesofCarrots and I both gave this the the green light; I restarted CI and this should get pushed in automatically once all the tests pass.

Thank you for contributing! Additionally, thank you for your patience with the initial delay before we took a look at the PR!

CriesofCarrots commented:
Thanks for your contribution and persistence, @gnapoli23 !

gnapoli23 commented:
@steviez @CriesofCarrots It was a pleasure, I hope I can give more help in the future 😄
